src/main/scala/is/hail/io/index/IndexWriter.scala

```diff
@@ -0,0 +1,165 @@
+    val fileOffset = trackedOS.bytesWritten
+    rootOffset = fileOffset
+
+    val firstKey = node.firstKey
+    val firstKeyOffset = node.firstOffset
+    val firstIndex = node.firstIndex
+
+    internalEncoder.writeByte(1)
+
+    val regionOffset = node.write(rvb)
+    internalEncoder.writeRegionValue(region, regionOffset)
+    internalEncoder.flush()
+
+    region.clear()
+    node.clear()
+
+    (fileOffset, firstIndex, firstKey, firstKeyOffset)
+  }
+
+  private def writeLeafNode(idx: Long): (Long, Any, Long) = {
+    val fileOffset = trackedOS.bytesWritten
+    val firstKey = leafNode.firstKey
+    val firstOffset = leafNode.firstOffset
+
+    leafEncoder.writeByte(0)
+
+    val regionOffset = leafNode.write(rvb, idx)
+    leafEncoder.writeRegionValue(region, regionOffset)
+    leafEncoder.flush()
+
+    region.clear()
+    leafNode.clear()
+
+    (fileOffset, firstKey, firstOffset)
+  }
+
+  private def write() {
+    def updateInternalNodes(level: Int, fileOffset: Long, firstIndex: Long, firstKey: Any, firstOffset: Long) {
+      if (level == internalNodes.size)
+        internalNodes += new InternalNodeBuilder(keyType)
+
+      val node = internalNodes(level)
+
+      if (node.size == 0) {
+        node.setFirstIndex(firstIndex)
+      }
+
+      node += (fileOffset, firstKey, firstOffset)
+
+      if (node.size == branchingFactor) {
+        val (fileOffset, firstIndex, firstKey, firstKeyOffset) = writeInternalNode(node)
+        updateInternalNodes(level + 1, fileOffset, firstIndex, firstKey, firstKeyOffset)
+      }
+    }
+
+    val idx = elementIdx - leafNode.size
+    val (leafFileOffset, leafFirstKey, leafFirstOffset) = writeLeafNode(idx)
+    updateInternalNodes(0, leafFileOffset, idx, leafFirstKey, leafFirstOffset)
+  }
+
+  private def flush() {
+    val nInternalLevels = math.max(IndexUtils.calcDepth(elementIdx, branchingFactor) - 1, 1) // ensure always one internal node
+
+    def flushInternalNodes(level: Int, fileOffset: Long, firstIndex: Long, firstKey: Any, firstOffset: Long) {
+      if (level != nInternalLevels) {
+        val node = internalNodes(level)
+
+        if (node.size == 0)
+          node.setFirstIndex(firstIndex)
+
+        node += (fileOffset, firstKey, firstOffset)
+
+        assert(node.size > 0 && node.size <= branchingFactor)
+
+        val (nextFileOffset, nextFirstIndex, nextFirstKey, nextFirstKeyOffset) = writeInternalNode(node)
+
+        flushInternalNodes(level + 1, nextFileOffset, nextFirstIndex, nextFirstKey, nextFirstKeyOffset)
+      }
+    }
+
+    if (leafNode.size > 0) {
+      write()
+    }
+    assert(internalNodes.size > 0)
+
+    var level = 0
+    while (level < nInternalLevels) {
+      val node = internalNodes(level)
+      if (node.size > 0) {
+        val (fileOffset, firstIndex, firstKey, firstKeyOffset) = writeInternalNode(node)
+        flushInternalNodes(level + 1, fileOffset, firstIndex, firstKey, firstKeyOffset)
```

Comment:
Hmm, I'm confused by this.  You have two loops: the loop here over level and the recursive loop in flushInternalNodes that also seems to be over level.  I feel like you only need one of these.

In particular, to flush, don't you need to write out the leaf, add that node to the first internal node, write it out, add that to the next internal node, etc.?

Reply:
I changed the code to hopefully make it clearer that there aren't actually two loops through the levels. The behavior of the code hasn't changed.